### PR TITLE
Binomial CDF scalar function added.

### PR DESCRIPTION
**Pull request by gopukrishnasIBM**
Resolves https://github.com/facebookincubator/velox/issues/4636

### CODE REVIEW

**Comment by mbasmanova** [comment]
@gopukrishnasIBM Thank you for the contribution. Please, review https://github.com/facebookincubator/velox/blob/main/CONTRIBUTING.md#prestos-sql-functions and https://github.com/facebookincubator/velox/blob/main/CODING_STYLE.md and update the PR to comply.

File: velox/functions/prestosql/Arithmetic.h [hunk 1]
```diff
@@ -495,5 +495,41 @@ struct TruncateFunction {
...
+    } else {
+      return (n * choose(n - 1, k - 1)) / k;
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE void
+  call(double& result, int64_t numOfTrials, double successProb, int64_t value) {
+    VELOX_USER_CHECK_GT(
+        numOfTrials, 0, "Number of trials must be greater than 0");
+    VELOX_USER_CHECK_GE(value, 0, "Value must be a positive integer");
```
Comment (aditi-pandit) [inline velox/functions/prestosql/Arithmetic.h:515]: Hmm.... I don't see this and the below check ("Number of trials must be greater than or equal to value") in the Presto java code.  Please can you clarify them.

File: velox/functions/prestosql/Arithmetic.h [hunk 2]
```diff
@@ -495,5 +495,41 @@ struct TruncateFunction {
...
+    VELOX_USER_CHECK_GE(
+        numOfTrials,
+        value,
+        "Number of trials must be greater than or equal to value");
+    VELOX_USER_CHECK(
+        (successProb >= 0) && (successProb <= 1),
+        "Success probability must be real value in [0, 1]");
+
+    double failureProb = 1 - successProb;
```
Comment (aditi-pandit) [inline velox/functions/prestosql/Arithmetic.h:524]: Is it possible to use binomial distribution from the boost libraries for this logic. e.g. https://www.boost.org/doc/libs/1_67_0/boost/math/distributions/binomial.hpp has a cdf funtion.

File: velox/functions/prestosql/tests/ArithmeticTest.cpp [hunk 3]
```diff
@@ -699,5 +699,34 @@ TEST_F(ArithmeticTest, truncate) {
...
 
+TEST_F(ArithmeticTest, binomialCDF) {
+  const auto binomialCDF = [&](std::optional<int64_t> numOfTrials,
+                               std::optional<double> successProb,
+                               std::optional<int64_t> value) {
+    return evaluateOnce<double>(
+        "binomial_cdf(c0, c1, c2)", numOfTrials, successProb, value);
+  };
+
+  EXPECT_EQ(binomialCDF(5, 0.5, 5), 1.0);
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/ArithmeticTest.cpp:710]: Please add tests with boundary values std::numeric_limits<double>::infinity(), std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::max(), std::numeric_limits<double>::min();

**Review by aditi-pandit — COMMENTED** [review]
Thanks @gopukrishnasIBM for adding this function.

**Comment by aditi-pandit** [comment]
Hi @gopukrishnasIBM : As per https://github.com/facebookincubator/velox/pull/4661 the cdf functions have been moved to Probability.h. Please can you refactor this PR to do so also.

File: velox/functions/prestosql/Arithmetic.h (hunk 1 above)
Comment (gopukrishnasIBM) [reply velox/functions/prestosql/Arithmetic.h]: @aditi-pandit Yes this check is missing in Java code, don't know why. Its clearly mentioned in presto binomial_cdf documentation ([binomial_cdf](https://prestodb.io/docs/current/functions/math.html?highlight=binomial#binomial_cdf)) as "numberOfTrials and value must be positive integers with numberOfTrials greater or equal to value".

File: velox/functions/prestosql/Arithmetic.h (hunk 2 above)
Comment (gopukrishnasIBM) [reply velox/functions/prestosql/Arithmetic.h]: Changed to boost library.

File: velox/functions/prestosql/tests/ArithmeticTest.cpp (hunk 3 above)
Comment (gopukrishnasIBM) [reply velox/functions/prestosql/tests/ArithmeticTest.cpp]: Added these test cases.

**Comment by gopukrishnasIBM** [comment]
Here is the output of Fuzzer test.
```
./velox_expression_fuzzer_test  --seed 123456 --duration_sec 60 --logtostderr=1 --minloglevel=0  --only binomial_cdf  --v=1 2>&1 | tee fuzzertest.out

I20230505 23:22:44.453264 16203878 ExpressionFuzzer.cpp:1275] ==============================> Done with iteration 8603
I20230505 23:22:44.453334 16203878 ExpressionFuzzer.cpp:1068] ==============================> Top 1 by number of rows processed
I20230505 23:22:44.453342 16203878 ExpressionFuzzer.cpp:1070] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230505 23:22:44.453369 16203878 ExpressionFuzzer.cpp:1074] binomial_cdf 17038 100.00% 533402
I20230505 23:22:44.453378 16203878 ExpressionFuzzer.cpp:1080] ==============================> Bottom 1 by number of rows processed
I20230505 23:22:44.453383 16203878 ExpressionFuzzer.cpp:1082] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230505 23:22:44.453406 16203878 ExpressionFuzzer.cpp:1087] binomial_cdf 17038 100.00% 533402
I20230505 23:22:44.453413 16203878 ExpressionFuzzer.cpp:1100] ==============================> All stats sorted by number of times the function was chosen
I20230505 23:22:44.453418 16203878 ExpressionFuzzer.cpp:1102] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230505 23:22:44.453445 16203878 ExpressionFuzzer.cpp:1106] binomial_cdf 17038 100.00% 533402
[==========] Running 0 tests from 0 test suites.
[==========] 0 tests from 0 test suites ran. (0 ms total)
[  PASSED  ] 0 tests.
```

**Comment by gopukrishnasIBM** [comment]
@aditi-pandit @mbasmanova Changes made, please review. Thanks.

File: velox/functions/prestosql/Arithmetic.h (hunk 1 above)
Comment (aditi-pandit) [reply velox/functions/prestosql/Arithmetic.h]: Thanks @gopukrishnasIBM. Did you try Presto SQL queries with such values in the invocation ? Would be interesting to see how Presto behaves. We can then evaluate if we want to add these errors.

File: velox/functions/prestosql/Probability.h [hunk 4]
```diff
@@ -43,5 +44,34 @@ struct BetaCDFFunction {
   }
 };
 
+template <typename T>
+struct BinomialCDFFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE void
+  call(double& result, int64_t numOfTrials, double successProb, int64_t value) {
+    constexpr int64_t kInf = std::numeric_limits<int64_t>::max();
```
[the remark this answers is not in the document]
Comment (gopukrishnasIBM) [reply velox/functions/prestosql/Probability.h:53]: @aditi-pandit Changed to static.

File: velox/functions/prestosql/Arithmetic.h (hunk 1 above)
Comment (gopukrishnasIBM) [reply velox/functions/prestosql/Arithmetic.h]: @aditi-pandit For this check, Java code returns '1.0' if the numberOfTrials is greater or equal to value. Can I implement the same logic here also ?
All other checks behave same.

Comment (aditi-pandit) [reply velox/functions/prestosql/Arithmetic.h]: @gopukrishnasIBM : Yes, I think we should make it consistent with Presto.

Comment (gopukrishnasIBM) [reply velox/functions/prestosql/Arithmetic.h]: @aditi-pandit Changes made to make it consistent with Presto.

File: velox/functions/prestosql/Probability.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (gopukrishnasIBM) [reply velox/functions/prestosql/Probability.h:53]: Will change that.

File: velox/functions/prestosql/Probability.h [hunk 5]
```diff
@@ -43,5 +44,30 @@ struct BetaCDFFunction {
...
+
+    if ((numOfTrials == kInf) || (value == kInf)) {
+      result = 0.0;
+      return;
+    }
+
+    VELOX_USER_CHECK_GT(
+        numOfTrials, 0, "Number of trials must be greater than 0");
+    VELOX_USER_CHECK_GE(value, 0, "Value must be a positive integer");
```
[the remark this answers is not in the document]
Comment (gopukrishnasIBM) [reply velox/functions/prestosql/Probability.h:62]: Added negative values.

File: velox/functions/prestosql/tests/ProbabilityTest.cpp [hunk 6]
```diff
@@ -26,6 +26,9 @@ constexpr double kInf = std::numeric_limits<double>::infinity();
 constexpr double kNan = std::numeric_limits<double>::quiet_NaN();
 constexpr double kDoubleMax = std::numeric_limits<double>::max();
 constexpr double kDoubleMin = std::numeric_limits<double>::min();
+constexpr int64_t kIntInf = std::numeric_limits<int64_t>::max();
```
[the remark this answers is not in the document]
Comment (gopukrishnasIBM) [reply velox/functions/prestosql/tests/ProbabilityTest.cpp:29]: Will change names.

[the remark this answers is not in the document]
Comment (gopukrishnasIBM) [reply velox/functions/prestosql/tests/ProbabilityTest.cpp:29]: @aditi-pandit No, the value returned by std::numeric_limits<int64_t>::infinity() is 0 and I could not use double infinity since my parameters are int64. So std::numeric_limits<int64_t>::max() is used as it returns a huge number (9223372036854775807).

File: velox/functions/prestosql/tests/ProbabilityTest.cpp [hunk 7]
```diff
@@ -69,5 +72,50 @@ TEST_F(ProbabilityTest, betaCDF) {
...
+        "binomial_cdf(c0, c1, c2)", numOfTrials, successProb, value);
+  };
+
+  EXPECT_EQ(binomialCDF(5, 0.5, 5), 1.0);
+  EXPECT_EQ(binomialCDF(5, 0.5, 0), 0.03125);
+  EXPECT_EQ(binomialCDF(3, 0.5, 1), 0.5);
+  EXPECT_EQ(binomialCDF(20, 1.0, 0), 0.0);
+  EXPECT_EQ(binomialCDF(20, 0.3, 6), 0.60800981220092398);
+  EXPECT_EQ(binomialCDF(200, 0.3, 60), 0.5348091761606989);
+  EXPECT_EQ(0.0, binomialCDF(kIntInf, 0.5, 5));
```
[the remark this answers is not in the document]
Comment (gopukrishnasIBM) [reply velox/functions/prestosql/tests/ProbabilityTest.cpp:89]: No reasons. Its a mistake. Changed to uniform invocations.

File: velox/functions/prestosql/tests/ProbabilityTest.cpp [hunk 8]
```diff
@@ -69,5 +72,51 @@ TEST_F(ProbabilityTest, betaCDF) {
...
+      binomialCDF(5, kBigIntInf, 3),
+      "successProbability must be in the interval [0, 1]");
+  VELOX_ASSERT_THROW(
+      binomialCDF(5, kBigIntMax, 3),
+      "successProbability must be in the interval [0, 1]");
+  VELOX_ASSERT_THROW(
+      binomialCDF(5, kNan, 3),
+      "successProbability must be in the interval [0, 1]");
+  VELOX_ASSERT_THROW(
+      binomialCDF(kNan, 0.5, 3), "numberOfTrials must be greater than 0");
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/ProbabilityTest.cpp:118]: Please also add some function invocations where more than one error condition are met. Please also add test cases where value < 0 and an error condition is true.

**Review by aditi-pandit — COMMENTED** [review]
This is looking quite good. Just have few more comments.

File: velox/functions/prestosql/tests/ProbabilityTest.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (aditi-pandit) [reply velox/functions/prestosql/tests/ProbabilityTest.cpp]: Hmm.... Then in that case there isn't really a distinction between the tests with kIntInf and kIntMax. Lets just leave the kIntMax tests around then.

File: velox/functions/prestosql/Probability.h [hunk 9]
```diff
@@ -43,5 +44,33 @@ struct BetaCDFFunction {
...
+  FOLLY_ALWAYS_INLINE void
+  call(double& result, int64_t numOfTrials, double successProb, int64_t value) {
+    static constexpr int64_t kInf = std::numeric_limits<int64_t>::max();
+
+    if (value < 0) {
+      result = 0.0;
+      return;
+    }
+
+    if ((numOfTrials == kInf) || (value == kInf)) {
```
[the remark this answers is not in the document]
Comment (gopukrishnasIBM) [reply velox/functions/prestosql/Probability.h]: @aditi-pandit Result is an error, not 0.0 when compared with presto java. Error logic moved ahead.

File: velox/functions/prestosql/tests/ProbabilityTest.cpp (hunk 8 above)
Comment (gopukrishnasIBM) [reply velox/functions/prestosql/tests/ProbabilityTest.cpp:151]: Added.

**Comment by gopukrishnasIBM** [comment]
@aditi-pandit Comments addressed. Please review.

**Review by aditi-pandit — APPROVED** [review]
This looks good. Thanks !

**Comment by aditi-pandit** [comment]
@Yuhta : Please can you take a look.

File: velox/functions/prestosql/Probability.h [hunk 10]
```diff
@@ -61,5 +62,29 @@ struct NormalCDFFunction {
...
+  call(double& result, int64_t numOfTrials, double successProb, int64_t value) {
+    static constexpr int64_t kInf = std::numeric_limits<int64_t>::max();
+
+    VELOX_USER_CHECK(
+        (successProb >= 0) && (successProb <= 1),
+        "successProbability must be in the interval [0, 1]");
+    VELOX_USER_CHECK_GT(
+        numOfTrials, 0, "numberOfTrials must be greater than 0");
+
+    if ((value < 0) || (numOfTrials == kInf) || (value == kInf)) {
```
Comment (Yuhta) [inline velox/functions/prestosql/Probability.h:79]: So this the same behavior as Presto?

File: velox/docs/functions/presto/math.rst [hunk 11]
```diff
@@ -245,8 +245,16 @@ Probability Functions
...
+
 .. function:: normal_cdf(mean, sd, value) -> double
 
     Compute the Normal cdf with given mean and standard deviation (sd): P(N < value; mean, sd).
     The mean and value must be real values and the standard deviation must be a real and
-    positive value (all of type DOUBLE).
\ No newline at end of file
+    positive value (all of type DOUBLE).
+
+    positive integers with numberOfTrials greater or equal to value.
```
[the remark this answers is not in the document]
Comment (gopukrishnasIBM) [reply velox/docs/functions/presto/math.rst]: @Yuhta Yes, fixed.

File: velox/functions/prestosql/Probability.h (hunk 10 above)
Comment (gopukrishnasIBM) [reply velox/functions/prestosql/Probability.h]: @Yuhta Compared with presto and corrected logic to match results.

File: velox/functions/prestosql/tests/ProbabilityTest.cpp [hunk 12]
```diff
@@ -108,5 +110,52 @@ TEST_F(ProbabilityTest, normalCDF) {
...
+      "successProbability must be in the interval [0, 1]");
+  VELOX_ASSERT_THROW(
+      binomialCDF(5, kNan, 3),
+      "successProbability must be in the interval [0, 1]");
+  VELOX_ASSERT_THROW(
+      binomialCDF(-1, 0.5, 2), "numberOfTrials must be greater than 0");
+  VELOX_ASSERT_THROW(
+      binomialCDF(kBigIntMin, 0.5, 1), "numberOfTrials must be greater than 0");
+  VELOX_ASSERT_THROW(
+      binomialCDF(kNan, 0.5, 3), "numberOfTrials must be greater than 0");
```
Comment (Yuhta) [inline velox/functions/prestosql/tests/ProbabilityTest.cpp:152]: We cannot cast from `kNan` to `long`, ASAN is failing on this line

Comment (gopukrishnasIBM) [reply velox/functions/prestosql/tests/ProbabilityTest.cpp]: @Yuhta Removed failing test case.
